src/Processors/Transforms/DistinctSortedTransform.cpp

```diff
@@ -143,16 +143,23 @@ ColumnRawPtrs DistinctSortedTransform::getClearingColumns(const ColumnRawPtrs &
 {
     ColumnRawPtrs clearing_hint_columns;
     clearing_hint_columns.reserve(description.size());
-    for (const auto & sort_column_description : description)
+    try
     {
-        const auto * sort_column_ptr = header.getByName(sort_column_description.column_name).column.get();
-        const auto it = std::find(key_columns.cbegin(), key_columns.cend(), sort_column_ptr);
-        if (it != key_columns.cend()) /// if found in key_columns
-            clearing_hint_columns.emplace_back(sort_column_ptr);
-        else
-            return clearing_hint_columns; /// We will use common prefix of sort description and requested DISTINCT key.
+        const size_t max_num_of_clearing_columns = std::min(description.size(), column_positions.size());
+        for (size_t i = 0; i < max_num_of_clearing_columns; ++i)
+        {
+            const auto pos = header.getPositionByName(description[i].column_name); // can throw
+            if (std::find(begin(column_positions), end(column_positions), pos) == column_positions.end())
+                break;
+
+            clearing_hint_columns.emplace_back(key_columns[pos]);
```

Comment:
Why we take `key_columns[pos]` while we checked `pos` in `column_positions`? Looks dangerous.

Reply:
Yes, looks like it should be `key_column[(column_positions^-1)[pos]]`

Reply:
Also, why do we do it for every chunk? We can just prepare an array of positions which are sorting key prefix fond in distinct keys.

Reply:
Position calculation is based on the header, which doesn't change. So, it's correct. But it's true, we don't need to calculate DISTINCT column positions, which form sort prefix of sort description, for every chunk, - only to build `clearing_columns` for every chunk is necessary, after `key_columns` is built from the `chunk`. I'll fix it

Overall, I'd prefer to not do much within this PR. It's just for the fix. `DistinctSortedTransform` refactoring looks to me as necessary thing, and I hope to do it in other PR, [see](https://github.com/ClickHouse/ClickHouse/pull/39538#issuecomment-1194026214)